### DISCUSSION TITLE
Type safe JOOQ layer

### DISCUSSION THREAD

**Discussion by aesteve** [jOOQ/jOOQ#15790] [Ideas]
Hello and thanks a lot for your work on JOOQ.

I'm not sure whether this is properly an "idea" but since it's not qualifying as Q&A per se, I chose this category, I hope it can be moved to the appropriate category if needed. Feel free to do so.

I was wondering whether it would be possible to push JOOQ even further in terms of static checking. Maybe an example could help clarifying my thoughts.

Say I have a `Person` (`Id` and `Name` columns) and a `PersonContact` table (`Address` and `Phone` column). Unless I'm mistaken, building a query the following way: `select().from(MySchema.PERSON).select(MySchema.PERSON_CONTACT.Address)` would compile and result in SQL being generated.

I was wondering if there could be a way to have columns be generic over their belonging table, something like `Column<Person>` and `Column<PersonContact>` so that when you build a query fragment (say `FromFragment<Person>`) you'd have a `select` method than only accepts `Column<Person>` (vararg potentially) as parameter. In case you built a JOIN, you'd have something like a `FromFragment2<Person, PersonContact>` and maybe 2 different select methods, that's not ideal but would still restrain from creating semantically incorrect. Same goes for `groupBy(Column<Person>)` .

I'm also wondering if this can be pushed even further for expressing conditions. If you build a query from Person only, the where condition shouldn't be built from `Column<PersonContact>` . Maybe using something like a `ConditionBuilder` could work here? I don't know.

That's all very abstract but I hope you get the idea of "using the compiler to help build more semantically correct queries" or rather, prevent users from building queries that don't make sense, like grouping on a field from a table that you haven't selected from (or joined with).

Do you think such a library can be built on top of JOOQ? Is this even possible at all (JOOQ or not)? Is this something you have in mind at some point? Maybe you know of a project that's actually doing that and I could use?

I tried to look for "type" "type safe" but not sure these are the most relevant keywords, feel free to redirect me to any existing discussion around this theme.

Thanks a lot.

**Comment by lukaseder** [comment]
Thanks for your comments.

I don't think this is possible without giving up on SQL syntax likeness, which is more important to jOOQ than any  additional "type safety." You'll find various query builders in other languages (especially Scala) who take this road. E.g. https://github.com/zio/zio-quill or https://scala-slick.org. You'll get a bit more type safety, yes, but from what I've seen from Slick user comments, the queries get really very hairy and hard to write/read, whereas with jOOQ, you can always pretty much 1:1 translate any SQL query to jOOQ.

I'm positive that a hypothetical "jOOQ for typescript" could do much better because its generics are so powerful,  but I don't think this kind of type safety is worth pursuing in Java.

**Reply by aesteve** [reply]
Thanks a lot for the pointers and explanations.
